src/Shared/Dictionary/SmallCapacityDictionary.cs

```diff
@@ -0,0 +1,760 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.AspNetCore.Internal.Dictionary
+{
+    /// <summary>
+    /// An <see cref="IDictionary{String, Object}"/> type to hold a small amount of items (4 or less in the common case).
+    /// </summary>
+    internal class SmallCapacityDictionary<TKey, TValue> : IDictionary<TKey, TValue>, IReadOnlyDictionary<TKey, TValue> where TKey : notnull
+    {
+        // Threshold for size of array to use.
+        private static readonly int DefaultArrayThreshold = 4;
+
+        internal KeyValuePair<TKey, TValue>[] _arrayStorage;
+        private int _count;
+        internal Dictionary<TKey, TValue>? _backup;
+        private IEqualityComparer<TKey> _comparer;
+
+        /// <summary>
+        /// Creates a new <see cref="SmallCapacityDictionary{TKey, TValue}"/> from the provided array.
+        /// The new instance will take ownership of the array, and may mutate it.
+        /// </summary>
+        /// <param name="items">The items array.</param>
+        /// <param name="comparer">Equality comparison.</param>
+        /// <returns>A new <see cref="SmallCapacityDictionary{TKey, TValue}"/>.</returns>
+        public static SmallCapacityDictionary<TKey, TValue> FromArray(KeyValuePair<TKey, TValue>[] items, IEqualityComparer<TKey>? comparer = null)
+        {
+            if (items == null)
+            {
+                throw new ArgumentNullException(nameof(items));
+            }
+
+            comparer = comparer ?? EqualityComparer<TKey>.Default;
+
+            if (items.Length > DefaultArrayThreshold)
+            {
+                // Use dictionary for large arrays.
+                var dict = new Dictionary<TKey, TValue>(items.Length, comparer);
+                foreach (var item in items)
+                {
+                    if (item.Key != null)
+                    {
+                        dict[item.Key] = item.Value;
+                    }
+                }
```

Comment:
```suggestion
                var dict = new Dictionary<TKey, TValue>(items, comparer);
```
?

Runtime-wise a difference is that if `item.Key == null` the libraries-code throws.